homeassistant/components/nissan_leaf.py

```diff
@@ -0,0 +1,338 @@
+"""
+Support for the Nissan Leaf Carwings/Nissan Connect API.
+
+Please note this is the pre-2018 API, which is still functional in the US.
+The old API should continue to work for the forseeable future.
+
+Documentation has not been created yet, here is an example configuration block:
+
+nissan_leaf:
+  username: "username"
+  password: "password"
+  nissan_connect: false
+  region: 'NE'
+  update_interval: 30
+  update_interval_charging: 15
+  update_interval_climate: 5
+  force_miles: true
+"""
+
+import logging
+from datetime import timedelta, datetime
+import time
+import urllib
+import asyncio
+import sys
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_USERNAME, CONF_PASSWORD
+from homeassistant.helpers.discovery import load_platform
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect, dispatcher_send)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import track_time_interval
+
+# Currently waiting on the lib author to review a PR
+# So using my fork for now.
+# https://github.com/jdhorne/pycarwings2/pull/28
+
+REQUIREMENTS = ['https://github.com/BenWoodford/pycarwings2/archive/master.zip'
+                '#pycarwings']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'nissan_leaf'
+DATA_LEAF = 'nissan_leaf_data'
+
+DATA_BATTERY = 'battery'
+DATA_LOCATION = 'location'
+DATA_CHARGING = 'charging'
+DATA_PLUGGED_IN = 'plugged_in'
+DATA_CLIMATE = 'climate'
+DATA_RANGE_AC = 'range_ac_on'
+DATA_RANGE_AC_OFF = 'range_ac_off'
+
+CONF_NCONNECT = 'nissan_connect'
+CONF_INTERVAL = 'update_interval'
+CONF_CHARGING_INTERVAL = 'update_interval_charging'
+CONF_CLIMATE_INTERVAL = 'update_interval_climate'
+CONF_REGION = 'region'
+CONF_VALID_REGIONS = ['NNA', 'NE', 'NCI', 'NMA', 'NML']
+CONF_FORCE_MILES = 'force_miles'
+DEFAULT_INTERVAL = 30
+DEFAULT_CHARGING_INTERVAL = 15
+DEFAULT_CLIMATE_INTERVAL = 5
+
+CHECK_INTERVAL = 10
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_REGION): vol.In(CONF_VALID_REGIONS),
+        vol.Optional(CONF_NCONNECT, default=True): cv.boolean,
+        vol.Optional(CONF_INTERVAL, default=DEFAULT_INTERVAL): cv.positive_int,

```

Comment:
Validate intervals with `vol.All(cv.time_period, cv.positive_timedelta)`, it will allow people to put in numbers but also "01:00". The validator converts all these different inputs to timedelta. You can get the value of the timedelta with `value.total_seconds()`.

Reply:
Why would we actually let anyone override the intervals? As developers we should pick the right interval and not bother our users with it. Especially we don't want to allow the users to be able to DDOS the vendor and us getting the blame.

Reply:
I also just saw the warning about emptying the 12v battery, yep definitely we should not allow users to pick their own interval. We should pick conservative intervals too.

Reply:
I've given it very conservative intervals but left it open to people making their own if they have it plugged in more. Alternatively we could look at decreasing the rate of polling if the drive train battery is on a low charge as that's when the 12V is more at risk (the drive train battery will top up the 12V intermittently)

Reply:
I like that idea very much.